src/dither_api.jl

```diff
@@ -84,59 +60,62 @@ end
 
 # Dispatch to binary dithering on grayscale images
 # when no color palette is provided
-function _dither!(
-    out::GenericGrayImage, img::GenericGrayImage, alg::AbstractDither; to_linear=false
-)
+function _dither(
+    ::Type{T}, img::GenericGrayImage, alg::AbstractDither; to_linear=false, kwargs...
+) where {T}
     to_linear && (img = srgb2linear.(img))
-    return binarydither!(alg, out, img)
+    index = binarydither(alg, img; kwargs...)
+    return IndirectArray(index, bwcolors(T))
 end
 
-# Dispatch to per-channel dithering on color images
-# when no color palette is provided
-function _dither!(
-    out::GenericImage{<:Color{<:Real,3},2},
+# Dispatch to per-channel dithering on color images when no color palette is provided
+function _dither(
+    ::Type{T},
     img::GenericImage{<:Color{<:Real,3},2},
     alg::AbstractDither;
     to_linear=false,
-)
+    kwargs...,
+) where {T}
     to_linear && (@warn "Skipping transformation `to_linear` when dithering color images.")
 
-    cvout = channelview(out)
-    cvimg = channelview(img)
-    for c in axes(cvout, 1)
-        # Note: the input `out` will be modified
-        # since the dithering algorithms modify the view of the channelview of `out`.
-        binarydither!(alg, view(cvout, c, :, :), view(cvimg, c, :, :))
+    cs = perchannelbinarycolors(T) # color scheme with binary respresentation
+    index = ones(Int, size(img)...) # allocate indices
+
+    for c in 1:3
+        channelindex = binarydither(alg, view(channelview(img), c, :, :), kwargs...)
+        index += 2^(3 - c) * (channelindex .- 1) # reconstruct "decimal" indices
```

Comment:
I'm sure there is a more better way...

Reply:
The binary dithering algorithms now output indices `1` = black, `2` = white for IndirectArrays with binary color schemes. For channel-wise dithering, these need to be converted into matching indices `1:2^3`. 
When using binary outputs, this can be done as `2^2 * r + 2^1 * g + 2^0 * b`, which is why I'm substracting 1.

Reply:
maybe `4*r + 2*g + b -7` and a comment would do the job. :)